pkg/util/workqueue/prometheus/prometheus.go

```diff
@@ -38,120 +38,156 @@ const (
-		Subsystem: WorkQueueSubsystem,
-		Name:      UnfinishedWorkKey,
-		Help: "How many seconds of work has done that " +
-			"is in progress and hasn't been observed by work_duration. Large " +
-			"values indicate stuck threads. One can deduce the number of stuck " +
-			"threads by observing the rate at which this increases.",
-		ConstLabels: prometheus.Labels{"name": name},
-	})
-	if err := prometheus.Register(unfinished); err != nil {
-		klog.Errorf("failed to register unfinished metric %v: %v", name, err)
-	}
-	return unfinished
+	return unfinished.WithLabelValues(name)
 }
 
 func (prometheusMetricsProvider) NewLongestRunningProcessorSecondsMetric(name string) workqueue.SettableGaugeMetric {
-	longestRunningProcessor := prometheus.NewGauge(prometheus.GaugeOpts{
-		Subsystem: WorkQueueSubsystem,
-		Name:      LongestRunningProcessorKey,
-		Help: "How many seconds has the longest running " +
-			"processor for workqueue been running.",
-		ConstLabels: prometheus.Labels{"name": name},
-	})
-	if err := prometheus.Register(longestRunningProcessor); err != nil {
-		klog.Errorf("failed to register unfinished metric %v: %v", name, err)
-	}
-	return longestRunningProcessor
+	return longestRunningProcessor.WithLabelValues(name)
 }
 
 func (prometheusMetricsProvider) NewRetriesMetric(name string) workqueue.CounterMetric {
-	retries := prometheus.NewCounter(prometheus.CounterOpts{
-		Subsystem:   WorkQueueSubsystem,
-		Name:        RetriesKey,
-		Help:        "Total number of retries handled by workqueue",
-		ConstLabels: prometheus.Labels{"name": name},
-	})
-	if err := prometheus.Register(retries); err != nil {
-		klog.Errorf("failed to register retries metric %v: %v", name, err)
-	}
-	return retries
+	return retries.WithLabelValues(name)
 }
 
 // TODO(danielqsj): Remove the following metrics, they are deprecated
+
+// mustRegister attempts to register the given collector with the given metric, and name
+// and returns the registered collector. The caller must use the returned collector
+// as it might point to a different instance of an already registered collector.
+func mustRegister(metric, name string, c prometheus.Collector) prometheus.Collector {
+	err := prometheus.Register(c)
+	if err == nil {
+		return c
+	}
+
+	if aerr, ok := err.(prometheus.AlreadyRegisteredError); ok {
+		klog.V(4).Infof("reusing already registered metric %v name %v", metric, name)
+		return aerr.ExistingCollector
+	}
+
+	// this should fail hard as this indicates a programmatic error, i.e.
+	// an invalid or duplicate metric descriptor,
+	// a previously registered descriptor with the same fqdn but different labels,
+	// or inconsistent label names or help strings for the same fqdn.
+	klog.Fatalf("failed to register metric %v name %v: %v", metric, name, err)
+	return nil
+}
+
 func (prometheusMetricsProvider) NewDeprecatedDepthMetric(name string) workqueue.GaugeMetric {
 	depth := prometheus.NewGauge(prometheus.GaugeOpts{
```

Comment:
nevermind, as discussed with @logicalhan we can go with simple singletons here and can omit mustRegister at all :+1: I will adapt the PR accordingly.

Reply:
Thanks!

Reply:
@logicalhan @brancz unfortunately declaring those deprecated metrics as singletons doesn't seem to work, as the subsystem is dynamic (sorry when I discussed this i was blind).

To omit the ugly mustRegister method, I added another commit which completely removes the deprecated metrics for good. @brancz the question is, if we can do this at this point.